File: examples/test_simple_race.cpp
// Copyright (c) 2017 Baidu Inc. All Rights Reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "simulator_interface.h"
#include "games/simple_race/simple_race_simulator.h"

using namespace simulator::simple_race;
using namespace simulator;

DECLARE_int32(context);

int main(int argc, char** argv) {
    gflags::ParseCommandLineFlags(&argc, &argv, true);

    auto race = std::make_shared<SimulatorInterface>("simple_race", false);
    race->reset_game();

    auto num_actions = race->get_num_actions();
    int act_rep = FLAGS_context;

    double reward = 0;
    double r = 0;
    bool show_screen = false;
    for (int i = 0; i < 100; i++) {
        auto game_over_str =
            GameSimulator::decode_game_over_code(race->game_over());

        auto state = race->get_state(r);

        StatePacket actions;
        actions.add_buffer_id("action", {util::get_rand_ind(num_actions)});
        r = race->take_actions(actions, act_rep, show_screen);

        if (game_over_str != "alive") {
            LOG(INFO) << "game over because of " + game_over_str;
            race->reset_game();
            continue;
        }

        reward += r;
        LOG(INFO) << r;
    }

    LOG(INFO) << "total reward " << reward;
    return 0;
}

File: python/examples/test_simple_race.py
#!/usr/bin/python

from py_simulator import Simulator
from random import randint

if __name__ == "__main__":
    options = {
        "pause_screen": False,
        "window_width": 480,
        "window_height": 480,
        "track_type": "straight",
        "track_width": 20.0,
        "track_length": 100.0,
        "track_radius": 30.0,
        "race_full_manouver": False,
        "random": False,
        "difficulty": "easy",
        "context": 1
    }
    sr = Simulator.create("simple_race", options)
    sr.reset_game()

    num_actions = sr.get_num_actions()
    act_rep = options["context"]

    reward = 0
    print("\033[93mUse show_screen() to see the game. " \
          + "However, X server must be used; otherwise the code crashes\033[0m")
    for i in range(100):
        game_over_str = sr.game_over()
        if game_over_str != "alive":
            print "game over because of ", game_over_str
            sr.reset_game()
            continue

#        sr.show_screen()

        states = sr.get_state()
        action = randint(0, num_actions - 1)
        r = sr.take_actions({"action": action}, act_rep)
        print r
        reward += r

    print "total reward ", reward
